File: python_files/raylib/space_shooter/player.py
import math

import pyray as p
from settings import MAP_SIZE, ROTATION_OFFSET


class Player:
    def __init__(
        self, spritesheet: p.Texture, sprite_dict: dict[str, p.Rectangle]
    ) -> None:
        self.current_sprite: str = "playerShip1_blue.png"

        self.texture: p.Texture = spritesheet
        self.sprite_data: dict[str, p.Rectangle] = sprite_dict

        self.scale: p.Vector2 = p.Vector2(0.7, 0.7)
        self.source: p.Rectangle = sprite_dict[self.current_sprite]
        self.dest: p.Rectangle = p.Rectangle(
            MAP_SIZE / 2.0,
            MAP_SIZE / 2.0,
            self.source.width * self.scale.x,
            self.source.height * self.scale.y,
        )
        self.origin: p.Vector2 = p.Vector2(
            self.dest.width / 2.0, self.dest.height / 2.0
        )
        self.rotation: float = ROTATION_OFFSET
        self.tint: p.Color = p.WHITE

        self.speed: float = 400.0
        self.move_direction: p.Vector2 = p.Vector2(0, 0)
        self.look_direction: p.Vector2 = p.Vector2(0, 0)

    def draw(self) -> None:
        p.draw_texture_pro(
            self.texture,
            self.source,
            self.dest,
            self.origin,
            self.rotation,
            self.tint,
        )

    def update(self, dt: float, mouse_pos_world: p.Vector2) -> None:
        # rotate towards mouse
        self.look_direction.x = mouse_pos_world.x - self.dest.x
        self.look_direction.y = mouse_pos_world.y - self.dest.y
        self.rotation = (
            math.degrees(
                math.atan2(self.look_direction.y, self.look_direction.x)
            )
            + ROTATION_OFFSET
        )

        # handle input
        self.move_direction.x = int(p.is_key_down(p.KeyboardKey.KEY_D)) - int(
            p.is_key_down(p.KeyboardKey.KEY_A)
        )
        self.move_direction.y = int(p.is_key_down(p.KeyboardKey.KEY_S)) - int(
            p.is_key_down(p.KeyboardKey.KEY_W)
        )

        if self.move_direction.x != 0 and self.move_direction.y != 0:
            self.move_direction = p.vector2_normalize(self.move_direction)

        # move
        self.dest.x += self.move_direction.x * self.speed * dt
        self.dest.y += self.move_direction.y * self.speed * dt

        # bounds
        self.dest.x = max(
            self.dest.width / 2.0,
            min(self.dest.x, MAP_SIZE - self.dest.width / 2.0),
        )
        self.dest.y = max(
            self.dest.height / 2.0,
            min(self.dest.y, MAP_SIZE - self.dest.height / 2.0),
        )

File: c_files/raylib/simple_space_survivors/src/player.c
#include "../include/player.h"
#include "../include/settings.h"

#include <assert.h>
#include <math.h>
#include <raymath.h>

static const float PLAYER_TEXTURE_SCALE = 0.7f;

void player_init(Player *player, const Texture2D spritesheet)
{
    assert(player);

    player->texture        = spritesheet;
    player->current_sprite = SPRITE_SHIP_A;
    player->source         = sprite_db[player->current_sprite];
    player->dest           = (Rectangle){MAP_SIZE / 2.0f, MAP_SIZE / 2.0f,
                                         player->source.width * PLAYER_TEXTURE_SCALE,
                                         player->source.height * PLAYER_TEXTURE_SCALE};
    player->origin =
        (Vector2){player->dest.width / 2.0f, player->dest.height / 2.0f};
    player->rotation = ROTATION_OFFSET;
    player->tint     = WHITE;

    player->speed          = 300.0f;
    player->look_direction = (Vector2){0, 0};
    player->move_direction = (Vector2){0, 0};

    player->max_hp = 100.0f;
    player->hp     = player->max_hp;
}

void player_draw(const Player *player)
{
    assert(player);

    DrawTexturePro(player->texture, player->source, player->dest,
                   player->origin, player->rotation, player->tint);
}

static void player_handle_input(Player *player)
{
    assert(player);

    player->move_direction.x = (int)IsKeyDown(KEY_D) - (int)IsKeyDown(KEY_A);
    player->move_direction.y = (int)IsKeyDown(KEY_S) - (int)IsKeyDown(KEY_W);

    player->move_direction = Vector2Normalize(player->move_direction);
}

static void player_rotate(Player *player, const Vector2 closest_target)
{
    assert(player);

    player->look_direction.x = closest_target.x - player->dest.x;
    player->look_direction.y = closest_target.y - player->dest.y;

    player->rotation =
        atan2f(player->look_direction.y, player->look_direction.x) * RAD2DEG +
        ROTATION_OFFSET;
}

static void player_move(Player *player, const float dt)
{
    assert(player);

    player->dest.x += player->move_direction.x * player->speed * dt;
    player->dest.y += player->move_direction.y * player->speed * dt;
}

static void player_bounds(Player *player)
{
    assert(player);

    player->dest.x =
        fmaxf(player->dest.width / 2.0f,
              fminf(player->dest.x, MAP_SIZE - player->dest.width / 2.0f));

    player->dest.y =
        fmaxf(player->dest.height / 2.0f,
              fminf(player->dest.y, MAP_SIZE - player->dest.height / 2.0f));
}

void player_update(Player *player, const float dt, const Vector2 closest_target)
{
    assert(player);

    player_handle_input(player);
    player_rotate(player, closest_target);
    player_move(player, dt);
    player_bounds(player);
}
